src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const arrayOfActions = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState;
+
+    switch (action.type) {
+      case 'addProperties':
+        newState = {
+          ...currentState, ...action.extraData,
+        };
+        break;
+
+      case 'removeProperties':
+        newState = { ...currentState };
+        action.keysToRemove.forEach(key => delete newState[key]);
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+
+      default:
+        throw new Error('Invalid action type');
+    }
+
+    arrayOfActions.push(newState);
+    currentState = newState;
```

Comment:
You can simplify your code a bit by avoiding `newState` alltogether, and just mutating `currentState` instead of it. After this, you can push the **copy** of current state to your array
```suggestion
    arrayOfActions.push({...currentState});
```